### PR TITLE
Use `ByteVector#toHex` in `CSRF` Middleware

### PR DESCRIPTION
**Pull request by danicheg**
`org.http4s.internal.{encodeHex,encodeHexString}` are seemingly too low-level to be a part of http4s, innit?

### CODE REVIEW

File: core/shared/src/main/scala/org/http4s/internal/package.scala [hunk 1]
```diff
@@ -49,12 +49,14 @@ package object internal extends InternalPlatform {
...
     */
+  @deprecated("Will be removed in 1.0.", "0.23.19")
   private[http4s] final def encodeHexString(data: Array[Byte]): String =
     new String(encodeHex(data))
 
   /** Encode a string to a Hexadecimal string representation
     * Adapted from apache commons Hex.encodeHex
     */
+  @deprecated("Will be removed in 1.0.", "0.23.19")
   private[http4s] final def encodeHex(data: Array[Byte]): Array[Char] = {
```
Comment (armanbilge) [inline core/shared/src/main/scala/org/http4s/internal/package.scala:60]: Yes! Always happy to see stuff in `internal` package go away!

**Review by armanbilge — APPROVED** [review]

**Review by rossabaker — COMMENTED** [review]
If I remember right, that code is several years old and someone worked real hard to squeeze the performance out of it.  Note the benchmark.  I'm all for getting rid of it's similar.  If it's not, I'm all for seeing if it can be improved in scodec.

If we do eliminate it, we don't need to maintain the deprecated benchmark for bincompat.

**Comment by danicheg** [comment]
That's a fair point about performance. But anyway, http4s should rely on scodec-like low-level libraries for doing things over bytes. 
For interested readers, it looks like the fs2 version is up to 4x times slower than the http4s one (or I sucked up with benchmarking, but it seems to be not).

**Comment by armanbilge** [comment]
Hmm, that's too bad. Here's the implementation in scodec, not surprised it's not very fast.

https://github.com/scodec/scodec-bits/blob/67ee8c23441d93270f3f98463be497bd013face9/core/shared/src/main/scala/scodec/bits/ByteVector.scala#L806-L819

At the very least the `StringBuilder` can be initialized with the right size. But still it's using an anonymous class to iterate over each byte ... probably not so good for JIT.

Maybe you can override for `AtArray` with a tight while loop.
https://github.com/scodec/scodec-bits/blob/67ee8c23441d93270f3f98463be497bd013face9/core/shared/src/main/scala/scodec/bits/ByteVector.scala#L1398

**Comment by danicheg** [comment]
> But still it's using an anonymous class to iterate over each byte ... probably not so good for JIT.

Yeah, it also got me stuck at first glance. Perhaps, there are reasons for iterating that way since it's a common design pattern within `scodec`.

**Comment by mpilquist** [comment]
Took a guess as to the difference in performance here: https://github.com/scodec/scodec-bits/pull/420

If there's still a significant gap, then next step would be inlining the pattern match from `foreachS` and having an optimized version for `AtView` like @armanbilge suggested. Diminishing returns of course, so I'd only head that way if the hex conversion is on a hot path.

**Review by armanbilge — APPROVED** [review]
I think we are good to go as of:
- https://github.com/http4s/http4s/pull/6998

Benchmarks:
- https://github.com/scodec/scodec-bits/pull/420#pullrequestreview-1310711951

**Comment by danicheg** [comment]
Yeah, I also think so. But Ross suggested removing the benchmark. By the way, it was partially ported in the `scodec`. So that refactoring really makes sense.

**Pull request merged**